rfcs/RFC-002-configurable-search-mode.md

```diff
@@ -0,0 +1,157 @@
+# Configurable Retrieval for RAG in Llama Stack
+
+** Authors:**
+
+* Red Hat: @varshaprasad96 @franciscojavierarceo
+
+## Summary
+
+This RFC proposes expanding the RAG capabilities in Llama Stack to support keyword search, hybrid search, and other
+retrieval strategies through a backend configuration.
+
+## Motivation
+
+The benefits of pre-retrieval optimization through indexing have been well
+studied ([1][rag-ref-1], [2][rag-ref-2], [3][rag-ref-3]) and it has been shown that keyword, vector, hybrid search, and other search strategies offer distinct benefits to
+information retrieval for RAG. Enabling Llama Stack to easily configure different search modes, while abstracting the implementation
+details offers significant value to Llama Stack users.
+
+## Scope
+
+### Goals:
+
+1. Create a design to support different search modes across supported databases.
+
+### Non-Goals:
+
+1. Multi-mode stacking: We will focus on a single selectable search mode per database.
+
+## Requirements:
+
+To support `keyword` based searches (and consequently hybrid search), a query string is required. Therefore the `query`
+method will need an additional parameter that is optional, we propose calling this new parameter `query_string`.
+
+There are at least three different implementation options available:
+1. Configure the Search Mode per application when registering the Vector Provider and applying the same mode for all
+queries in the application.
+2. Configure the Search Mode per query when querying the Vector Provider and allowing for different search modes for each
+query in the application.
+3. Create a new `keyword_search_io` provider and create a separate implementation for each provider.
+
+A brief review of the pros and cons of the three options are provided in the table below.
+
+### Implementation Evaluation
+
+| Option                                               | Pros                                                                                                            | Cons                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
```

Comment:
Can this be formatted as three bullet lists a la

```
* Option 1: ...
  - Pros
    1. ...
    2. ...
  - Cons
    1. ...
    2. ...
* Options 2: ...
 ....
```

I think this way it's easier to digest and comment on in MD source and rendered version.

Reply:
Sure, if that's easier I'll reformat.